### ISSUE TITLE
Actual dataset classification is less than labels

### ISSUE THREAD

**Issue by wting861006** [AlexeyAB/darknet#3280]
@AlexeyAB 
Hi,I want to train a model with 4000 categories.But only have 500 classification dataset in this time while I will add more categories in the future.Can you give me some advice now?

I use yolov3-spp.cfg:
batch=16
subdivisions=8
width=608
height=608
learning_rate=0.0001
classes=4000 in each of 3 [yolo]-layers
filters=12015 in the 3 [convolutional] before each [yolo] layer

**Comment by AlexeyAB** [comment]
@wting861006 Hi,

Train with
```
classes=4000 in each of 3 [yolo]-layers
filters=12015 in the 3 [convolutional] before each [yolo] layer
```

And when you add new objects, then continue training with `old_dataset + new_dataset`

**Comment by wting861006** [comment]
@wting861006 Hi,
This is my train log,I think the result is not good.

Last accuracy mAP@0.5 = 1.62 %, best = 1.62 % 
 146880: 121.146484, 128.400024 avg loss, 0.000004 rate, 5.736079 seconds, 4700160 images
Loaded: 0.000046 seconds
Region 89 Avg IOU: 0.349577, Class: 0.028772, Obj: 0.090707, No Obj: 0.047867, .5R: 0.220779, .75R: 0.045455,  count: 154
Region 89 Avg IOU: 0.330239, Class: 0.013179, Obj: 0.100670, No Obj: 0.042956, .5R: 0.230769, .75R: 0.153846,  count: 13
Region 89 Avg IOU: 0.310353, Class: 0.021537, Obj: 0.111669, No Obj: 0.046967, .5R: 0.118644, .75R: 0.033898,  count: 59
Region 89 Avg IOU: 0.345305, Class: 0.021432, Obj: 0.089220, No Obj: 0.049600, .5R: 0.178808, .75R: 0.039735,  count: 151
Region 101 Avg IOU: 0.740141, Class: 0.464832, Obj: 0.515111, No Obj: 0.021187, .5R: 0.948718, .75R: 0.589744,  count: 78
Region 101 Avg IOU: 0.726763, Class: 0.366962, Obj: 0.455059, No Obj: 0.021550, .5R: 0.902597, .75R: 0.538961,  count: 154
Region 101 Avg IOU: 0.767792, Class: 0.403637, Obj: 0.515740, No Obj: 0.021141, .5R: 0.992537, .75R: 0.597015,  count: 134
Region 101 Avg IOU: 0.725869, Class: 0.387593, Obj: 0.499640, No Obj: 0.025871, .5R: 0.897959, .75R: 0.545918,  count: 196
Region 113 Avg IOU: 0.741401, Class: 0.100860, Obj: 0.337994, No Obj: 0.002889, .5R: 0.924051, .75R: 0.594937,  count: 79
Region 113 Avg IOU: 0.778043, Class: 0.086826, Obj: 0.476780, No Obj: 0.005829, .5R: 0.981707, .75R: 0.682927,  count: 164
Region 113 Avg IOU: 0.783053, Class: 0.074814, Obj: 0.530327, No Obj: 0.012129, .5R: 0.995305, .75R: 0.708920,  count: 213
Region 113 Avg IOU: 0.775443, Class: 0.074055, Obj: 0.469577, No Obj: 0.006909, .5R: 0.988571, .75R: 0.662857,  count: 175
Region 89 Avg IOU: 0.289506, Class: 0.024213, Obj: 0.110561, No Obj: 0.049418, .5R: 0.148352, .75R: 0.010989,  count: 182
Region 89 Avg IOU: 0.331488, Class: 0.023914, Obj: 0.071593, No Obj: 0.053193, .5R: 0.175627, .75R: 0.021505,  count: 279
Region 89 Avg IOU: 0.298995, Class: 0.020024, Obj: 0.053106, No Obj: 0.045617, .5R: 0.191057, .75R: 0.012195,  count: 246
Region 89 Avg IOU: 0.339938, Class: 0.020962, Obj: 0.094843, No Obj: 0.052772, .5R: 0.220339, .75R: 0.016949,  count: 59
Region 101 Avg IOU: 0.748451, Class: 0.456646, Obj: 0.534611, No Obj: 0.023076, .5R: 0.953125, .75R: 0.625000,  count: 128
Region 101 Avg IOU: 0.772788, Class: 0.411502, Obj: 0.393684, No Obj: 0.018022, .5R: 0.969388, .75R: 0.704082,  count: 98
Region 101 Avg IOU: 0.628350, Class: 0.380417, Obj: 0.502259, No Obj: 0.025891, .5R: 0.786047, .75R: 0.306977,  count: 215
Region 101 Avg IOU: 0.633628, Class: 0.325094, Obj: 0.583879, No Obj: 0.032071, .5R: 0.756923, .75R: 0.400000,  count: 325
Region 113 Avg IOU: 0.690577, Class: 0.084187, Obj: 0.324832, No Obj: 0.004263, .5R: 0.853658, .75R: 0.560976,  count: 41
Region 113 Avg IOU: 0.775859, Class: 0.103498, Obj: 0.255589, No Obj: 0.001912, .5R: 0.977778, .75R: 0.733333,  count: 45
Region 113 Avg IOU: 0.690974, Class: 0.186987, Obj: 0.171294, No Obj: 0.001211, .5R: 0.888889, .75R: 0.444444,  count: 18
Region 113 Avg IOU: 0.734962, Class: 0.072602, Obj: 0.308163, No Obj: 0.004054, .5R: 0.921053, .75R: 0.605263,  count: 114
Region 89 Avg IOU: 0.223508, Class: 0.020762, Obj: 0.114924, No Obj: 0.047944, .5R: 0.062500, .75R: 0.000000,  count: 32
Region 89 Avg IOU: 0.309066, Class: 0.020677, Obj: 0.091607, No Obj: 0.050668, .5R: 0.155963, .75R: 0.018349,  count: 218
Region 89 Avg IOU: 0.304179, Class: 0.019465, Obj: 0.109285, No Obj: 0.053657, .5R: 0.088889, .75R: 0.016667,  count: 180
Region 101 Avg IOU: 0.553184, Class: 0.381072, Obj: 0.613604, No Obj: 0.021469, .5R: 0.636986, .75R: 0.246575,  count: 146
Region 89 Avg IOU: 0.399214, Class: 0.023076, Obj: 0.108758, No Obj: 0.052571, .5R: 0.317073, .75R: 0.048780,  count: 123
Region 101 Avg IOU: 0.777364, Class: 0.344239, Obj: 0.508321, No Obj: 0.022056, .5R: 1.000000, .75R: 0.681818,  count: 132
Region 101 Avg IOU: 0.734856, Class: 0.376078, Obj: 0.411875, No Obj: 0.024239, .5R: 0.942308, .75R: 0.519231,  count: 104
Region 101 Avg IOU: 0.773506, Class: 0.445470, Obj: 0.570496, No Obj: 0.032258, .5R: 0.984848, .75R: 0.651515,  count: 198
Region 113 Avg IOU: 0.734140, Class: 0.061254, Obj: 0.502914, No Obj: 0.009304, .5R: 0.941177, .75R: 0.546219,  count: 238
Region 113 Avg IOU: 0.711879, Class: 0.074860, Obj: 0.319056, No Obj: 0.003061, .5R: 0.932203, .75R: 0.457627,  count: 59
Region 113 Avg IOU: 0.750571, Class: 0.060877, Obj: 0.388776, No Obj: 0.006074, .5R: 0.958115, .75R: 0.612565,  count: 191
Region 113 Avg IOU: 0.764838, Class: 0.073668, Obj: 0.403309, No Obj: 0.006096, .5R: 0.973214, .75R: 0.633929,  count: 112
Region 89 Avg IOU: 0.294927, Class: 0.026142, Obj: 0.100348, No Obj: 0.045424, .5R: 0.228070, .75R: 0.000000,  count: 57
Region 89 Avg IOU: 0.284458, Class: 0.010538, Obj: 0.128510, No Obj: 0.047431, .5R: 0.142857, .75R: 0.000000,  count: 7
Region 101 Avg IOU: 0.801943, Class: 0.490723, Obj: 0.535598, No Obj: 0.019100, .5R: 0.989691, .75R: 0.835052,  count: 97
Region 89 Avg IOU: 0.315847, Class: 0.027445, Obj: 0.103298, No Obj: 0.045716, .5R: 0.182540, .75R: 0.007937,  count: 126
Region 101 Avg IOU: 0.777160, Class: 0.624860, Obj: 0.471641, No Obj: 0.011672, .5R: 1.000000, .75R: 0.757576,  count: 33
Region 89 Avg IOU: 0.224311, Class: 0.016261, Obj: 0.114754, No Obj: 0.042528, .5R: 0.000000, .75R: 0.000000,  count: 38
Region 101 Avg IOU: 0.776059, Class: 0.463738, Obj: 0.487168, No Obj: 0.023698, .5R: 0.947368, .75R: 0.714286,  count: 133
Region 101 Avg IOU: 0.755165, Class: 0.468994, Obj: 0.724035, No Obj: 0.040469, .5R: 0.989051, .75R: 0.572993,  count: 274
Region 113 Avg IOU: 0.769180, Class: 0.077138, Obj: 0.547530, No Obj: 0.009515, .5R: 0.969565, .75R: 0.678261,  count: 230
Region 113 Avg IOU: 0.754623, Class: 0.062018, Obj: 0.645483, No Obj: 0.015956, .5R: 0.980687, .75R: 0.564378,  count: 466
Region 113 Avg IOU: 0.728218, Class: 0.089087, Obj: 0.450022, No Obj: 0.004633, .5R: 0.932584, .75R: 0.561798,  count: 89
Region 113 Avg IOU: 0.740749, Class: 0.065625, Obj: 0.520724, No Obj: 0.008655, .5R: 0.935065, .75R: 0.603896,  count: 154

 for thresh = 0.25, precision = 0.71, recall = 0.18, F1-score = 0.29 
 for thresh = 0.25, TP = 458, FP = 184, FN = 2064, average IoU = 54.94 % 

 IoU threshold = 50 %, used Area-Under-Curve for each unique Recall 
 mean average precision (mAP@0.50) = 0.016151, or 1.62 % 
Total Detection Time: 20.000000 Seconds

Set -points flag:
 `-points 101` for MS COCO 
 `-points 11` for PascalVOC 2007 (uncomment `difficult` in voc.data) 
 `-points 0` (AUC) for ImageNet, PascalVOC 2010-2012, your custom dataset

 mean_average_precision (mAP@0.5) = 0.016151 
New best mAP!
Saving weights to /home/wangting/Chinese/backup/yolov3-spp_best.weights

**Comment by AlexeyAB** [comment]
It's because you don't use the most of classes.

* How many classes do you actually use?

* How many classes did you set in yolo-layer?

**Comment by wting861006** [comment]
> It's because you don't use the most of classes.
> 
> * How many classes do you actually use?
> * How many classes did you set in yolo-layer?

@AlexeyAB Hi，Thanks for you reply.
I use actually about 100 classes.In the yolo-layer I set 4000 classes.

**Comment by AlexeyAB** [comment]
So your mAP is calculated as  `mAP = (AP1 + AP2 + ... AP100 + 0 + 0 + ... + 0) / 4000`

So you can think that actually your `mAP =  1.62 % * 4000 / 100 = 64.8%` it is a good mAP.
